client/asset/btc/electrum_client.go

```diff
@@ -0,0 +1,1289 @@
+		addr, tlsConfig, err := ew.connInfo(host)
+		if err != nil {
+			return "", nil, fmt.Errorf("no suitable address for host %q: %w", host, err)
+		}
+		srvOpts = &electrum.ConnectOpts{
+			// TorProxy: TODO
+			TLSConfig: tlsConfig, // may be nil if not ssl host
+		}
+		return addr, srvOpts, nil
+	}
+
+	info, err := ew.wallet.GetInfo() // also initial connectivity test with the external wallet
+	if err != nil {
+		return err
+	}
+	if !info.Connected || info.Server == "" {
+		return errors.New("Electrum wallet has no server connections")
+	}
+
+	// Determine if segwit expectation is met. Request and decode an address,
+	// then compare with the segwit config field.
+	addr, err := ew.wallet.GetUnusedAddress()
+	if err != nil {
+		return err
+	}
+	address, err := ew.decodeAddr(addr, ew.chainParams)
+	if err != nil {
+		return err
+	}
+	_, segwit := address.(interface {
+		WitnessVersion() byte
+	})
+	if segwit != ew.segwit {
+		return fmt.Errorf("segwit expectation not met: wanted segwit = %v", ew.segwit)
+	}
+
+	addr, srvOpts, err := connInfo(info.Server)
+	if err != nil {
+		return fmt.Errorf("no suitable address for host %v: %w", info.Server, err)
+	}
+	chain, err := electrum.ConnectServer(ctx, addr, srvOpts)
+	if err != nil {
+		return err // maybe just try a different one if it doesn't allow multiple conns
+	}
+	ew.log.Infof("Now connected to electrum server %v.", addr)
+	ew.chain = &electrumNetworkManager{electrumNetworkClient: chain}
+
+	// Start a goroutine to keep the chain client alive and on the same
+	// ElectrumX server as the external Electrum wallet if possible.
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		defer ew.chain.Shutdown()
+		lastWalletServer := info.Server
+
+		failing := make(map[string]int)
+		const maxFails = 8
+
+		ticker := time.NewTicker(6 * time.Second) // to keep wallet and chain client on same server
+		defer ticker.Stop()
+
+		for {
+			var walletCheck bool
+			select {
+			case <-ew.chain.Done():
+				ew.log.Warnf("Electrum server connection lost. Reconnecting in 5 seconds...")
+				select {
+				case <-time.After(5 * time.Second):
+				case <-ctx.Done():
+					return
+				}
+
+			case <-ticker.C: // just checking with wallet for changes
+				walletCheck = true
+
+			case <-ctx.Done():
+				return
+			}
+
+			info, err := ew.wallet.GetInfo()
+			if err != nil {
+				ew.log.Errorf("Electrum wallet getinfo failed: %v", err)
+				continue
+			}
+			if walletCheck { // just checking if wallet's server changed
+				if lastWalletServer == info.Server {
+					continue // no change
+				}
+				delete(failing, info.Server) // clean slate now that wallet has just gotten on it
+				ew.log.Infof("Electrum wallet changed server to %v", info.Server)
+			}
+			lastWalletServer = info.Server
+
+			tryAddr := info.Server
+			if fails := failing[tryAddr]; fails > maxFails {
+				ew.log.Warnf("Server %q has failed to connect %d times. Trying a random one...", tryAddr, fails)
+				tryAddr = "" // try a random one instead
```

Comment:
I would've thought it would do the same thing each time if the map is exactly the same. I think you should at least keep track of which ones failed when doing `failing[tryAddr]++` though, instead of just saying that picking a random one failed.

Reply:
Probably off topic,

> as long as "not specified" means random, but that's probably a bad idea

Found it curious, that 1st element of the map gets chosen roughly 50% of the time (using range over map): https://go.dev/play/p/s1wrj_UY41Q?v=gotip

Reply:
That's quite the bias.  Yeah, definitely not an approach to take. Was being exceedingly lazy while hacking on this logic